primitives/header-chain/src/justification.rs

```diff
@@ -63,119 +87,136 @@ pub fn verify_justification<Header: HeaderT>(
 		return Err(Error::InvalidJustificationTarget);
 	}
 
-	// Validate commit of the justification. Note that `validate_commit()` assumes that all
-	// signatures are valid. We'll check the validity of the signatures later since they're more
-	// resource intensive to verify.
-	let ancestry_chain = AncestryChain::new(&justification.votes_ancestries);
-	match finality_grandpa::validate_commit(&justification.commit, authorities_set, &ancestry_chain) {
-		Ok(ref result) if result.ghost().is_some() => {}
-		_ => return Err(Error::InvalidJustificationCommit),
-	}
-
-	// Now that we know that the commit is correct, check authorities signatures
-	let mut buf = Vec::new();
-	let mut visited_hashes = BTreeSet::new();
+	let mut chain = AncestryChain::new(&justification.votes_ancestries);
+	let mut signature_buffer = Vec::new();
+	let mut votes = BTreeSet::new();
+	let mut cumulative_weight = 0u64;
 	for signed in &justification.commit.precommits {
+		// authority must be in the set
+		let authority_info = match authorities_set.get(&signed.id) {
+			Some(authority_info) => authority_info,
+			None => {
+				// just ignore precommit from unknown authority as `finality_grandpa::import_precommit` does
+				continue;
+			}
+		};
+		// check if authority has already voted in the same round.
+		//
+		// there's a lot of code in `validate_commit` and `import_precommit` functions inside
+		// `finality-grandpa` crate (mostly related to reporing equivocations). But the only thing that we
+		// care about is that only first vote from the authority is accepted
+		if !votes.insert(signed.id.clone()) {
+			continue;
+		}
+
+		// everything below this line can't just `continue`, because state is already altered
+
+		// all precommits must be for block higher than the target
+		if signed.precommit.target_number < justification.commit.target_number {
+			return Err(Error::PrecommitTargetHasLowerNumberThanCommitTarget);
+		}
+		// all precommits must be for target block descendents
+		let is_descendant = chain.is_ancestor(&justification.commit.target_hash, &signed.precommit.target_hash);
+		if !is_descendant {
+			return Err(Error::PrecommitTargetIsNotCommitDescendant);
+		}
+		// since we know now that the precommit target is the descendant of the justification target,
+		// we may increase 'weight' of the justification target
+		//
+		// there's a lot of code in the `VoteGraph::insert` method inside `finality-grandpa` crate,
+		// but in the end it is only used to find GHOST, which we don't care about. The only thing
+		// that we care about is that the justification target has enough weight
+		cumulative_weight = cumulative_weight.checked_add(authority_info.weight().0.into()).expect(
+			"sum of weights of ALL authorities is expected not to overflow - this is guaranteed by\
+				existence of VoterSet;\
+				the order of loop conditions guarantees that we can account vote from same authority\
+				multiple times;\
+				thus we'll never overflow the u64::MAX;\
```

Comment:
(1) total weight of all authorities in the `VoterSet` can't overflow and (2) we only accept vote from single authority (it is ignored above) => it can't overflow.